include/ensmallen_bits/ada_belief/ada_belief_update.hpp

```diff
@@ -0,0 +1,161 @@
+   */
+  AdaBeliefUpdate(const double epsilon = 1e-12,
+                  const double beta1 = 0.9,
+                  const double beta2 = 0.999) :
+    epsilon(epsilon),
+    beta1(beta1),
+    beta2(beta2),
+    iteration(0)
+  {
+    // Nothing to do.
+  }
+
+  //! Get the value for numerical stability.
+  double Epsilon() const { return epsilon; }
+  //! Modify the value used for numerical stability.
+  double& Epsilon() { return epsilon; }
+
+  //! Get the exponential decay rate for the 1st moment estimates.
+  double Beta1() const { return beta1; }
+  //! Modify the exponential decay rate for the 1st moment estimates.
+  double& Beta1() { return beta1; }
+
+  //! Get the exponential decay rate for the 2nd moment estimates.
+  double Beta2() const { return beta2; }
+  //! Modify the exponential decay rate for the 2nd moment estimates.
+  double& Beta2() { return beta2; }
+
+  //! Get the current iteration number.
+  size_t Iteration() const { return iteration; }
+  //! Modify the current iteration number.
+  size_t& Iteration() { return iteration; }
+
+  /**
+   * The UpdatePolicyType policy classes must contain an internal 'Policy'
+   * template class with two template arguments: MatType and GradType.  This is
+   * instantiated at the start of the optimization, and holds parameters
+   * specific to an individual optimization.
+   */
+  template<typename MatType, typename GradType>
+  class Policy
+  {
+   public:
+    /**
+     * This constructor is called by the SGD Optimize() method before the start
+     * of the iteration update process.
+     *
+     * @param parent AdaBeliefUpdate object.
+     * @param rows Number of rows in the gradient matrix.
+     * @param cols Number of columns in the gradient matrix.
+     */
+    Policy(AdaBeliefUpdate& parent, const size_t rows, const size_t cols) :
+        parent(parent)
+    {
+      m.zeros(rows, cols);
+      s.zeros(rows, cols);
+    }
+
+    /**
+     * Update step for AdaBelief.
+     *
+     * @param iterate Parameters that minimize the function.
+     * @param stepSize Step size to be used for the given iteration.
+     * @param gradient The gradient matrix.
+     */
+    void Update(MatType& iterate,
+                const double stepSize,
+                const GradType& gradient)
+    {
+      // Increment the iteration counter variable.
+      ++parent.iteration;
+
+      m *= parent.beta1;
+      m += (1 - parent.beta1) * gradient;
+
+      s *= parent.beta2;
+      s += (1 - parent.beta2) * arma::pow(gradient - m, 2.0);
+
+      const double biasCorrection1 = 1.0 - std::pow(parent.beta1,
+          parent.iteration);
+      const double biasCorrection2 = 1.0 - std::pow(parent.beta2,
+          parent.iteration);
+
+      // And update the iterate.
+      iterate -= (stepSize / biasCorrection1) * m /
+          (arma::sqrt(s + parent.epsilon) / (std::sqrt(biasCorrection2) +
+          parent.epsilon));
```

Comment:
This looks a little different than the `AdamUpdate` implementation, but it does seem to be correct to me.

Reply:
I used the Adam update approximation as a basis, but the difference is so small, that I think it makes sense to just follow the paper, to avoid any extra confusion.